k8s/upgrade-job/src/events/event_recorder.rs

```diff
@@ -179,38 +180,73 @@ impl EventRecorderBuilder {
                     secondary: event.secondary.clone(),
                 };
                 if let Err(error) = event_handler.publish(event).await.context(EventPublish) {
-                    tracing::error!(?error, "Failed to publish event");
+                    error!(%error);
                 }
 
                 select! {
                     _ = sleep(Duration::from_secs(1200)) => {}
                     event = rx.recv() => { current_event = event }
                 }
             }
-        })
-        .abort_handle();
 
-        Ok(EventRecorder {
-            event_loop_killswitch: aborter,
-            event_sender: tx,
-            from_version,
-            to_version,
-        })
+            // Acknowledgement for exit of event loop, i.e. all events published, nothing left
+            // behind in the channel.
+            if let Err(error) = ack_sender
+                .send(())
+                .await
+                .context(EventFinalAcknowledgementSend)
+            {
+                error!(%error);
+            }
+        });
+
+        Ok(event_recorder)
     }
 }
 
 /// This is a wrapper around a kube::runtime::events::Recorder.
 pub(crate) struct EventRecorder {
-    event_loop_killswitch: task::AbortHandle,
-    event_sender: mpsc::UnboundedSender<Event>,
+    event_sender: Option<mpsc::UnboundedSender<Event>>,
+    ack_sender: Option<mpsc::Sender<()>>,
+    ack_receiver: mpsc::Receiver<()>,
     from_version: String,
     to_version: String,
 }
 
 impl Drop for EventRecorder {
     fn drop(&mut self) {
         // Cancel the tokio task.
-        self.event_loop_killswitch.abort();
+        if self.event_sender.is_some() {
+            let sender = self.event_sender.clone().unwrap();
+            self.event_sender = None;
+            drop(sender);
+        }
+
+        let mut retry_count: u32 = 0;
+        let retry_duration = Duration::from_secs(1);
+        while retry_count < 10 {
+            match self
+                .ack_receiver
+                .try_recv()
+                .context(EventFinalAcknowledgementRecv)
+            {
+                Ok(_) => {
+                    if self.ack_sender.is_some() {
+                        let sender = self.ack_sender.clone().unwrap();
+                        self.ack_sender = None;
+                        drop(sender);
+                    }
+                    return;
+                }
+                Err(warning) => {
+                    warn!(%warning);
+                    std::thread::sleep(retry_duration);
```

Comment:
You shouldn't block like this.. it'll prevent async code on the same executor from running.

Reply:
You probably shouldn't write this kind of code in the drop, rust has no async drop yet sadly...

Reply:
Moved it to a different method, no longer on the drop.